### PR TITLE
Now migration callouts

### LINKED ISSUE #6018: Now Migration: In-body Callout updates
- Update the (`NewsFeature`) https://github.com/uiowa/uiowa/blob/7428cb6e5cef6acc82f4e26e7c92c184641dc94d/docroot/sites/now.uiowa.edu/modules/now_migrate/src/Plugin/migrate/source/Achievement.php#L195-L233
   and (`Achievement`) https://github.com/uiowa/uiowa/blob/7428cb6e5cef6acc82f4e26e7c92c184641dc94d/docroot/sites/now.uiowa.edu/modules/now_migrate/src/Plugin/migrate/source/NewsFeature.php#L338-L376 Both need the same updates to map the source callout to the new callout structure.
   - Adjust hardcoded defaults to determined size (small) and background (gray)
   - Test in the migration and check that the hardcoded callout construction replacements work well with various different bits of content from the source
   - Can find a query adjustment to ease the testing burden of this and pull in only nodes which contain callouts

### PR DESCRIPTION
**Pull request by GaryRidgway**
This is to track and test now migrations for achievement and news feature callouts.

Sync the site in dev desktop (uiowa703)
`git checkout in-body-callout-functionality__now-migration && git pull && ddev blt ds --site=now.uiowa.edu && ddev drush @now.local en now_migrate -y && ddev drush @now.local uli /admin/config/sitenow/sitenow-migrate`
Complete the form

I find the following command helpful:
`ddev drush @now.local mrs now_news_feature && ddev drush @now.local mr now_news_feature && ddev drush @now.local mim now_news_feature --limit=100`

replace `now_news_feature` with `now_achievement` and test again. If everything looks good, yeehaw.

### CODE REVIEW

File: docroot/sites/now.uiowa.edu/modules/now_migrate/src/Plugin/migrate/source/NewsFeature.php [hunk 1]
```diff
@@ -339,18 +340,23 @@ private function captionReplace($match) {
    * Helper function to update a callout during a preg_replace.
    */
   private function calloutReplace($match) {
+    $nid = \Drupal::routeMatch()->getRawParameter('node');
```
Comment (cory-skeers) [inline docroot/sites/now.uiowa.edu/modules/now_migrate/src/Plugin/migrate/source/NewsFeature.php]: Is this a leftover from something else? It doesn't look like it gets used anywhere.

Comment (GaryRidgway) [reply docroot/sites/now.uiowa.edu/modules/now_migrate/src/Plugin/migrate/source/NewsFeature.php]: 'Tis! good catch

**Comment by cory-skeers** [comment]
Pulling down to test

**Comment by cory-skeers** [comment]
There's this unfortunate intersection with the featured video and callout:
https://now.uiowa.ddev.site/news/2015/03/student-experience-grant-young
https://now.uiowa.edu/2015/03/student-experience-grant-young

image

The logic that places the video in the body checks for any images within the first 700 characters of the body--if there is one, it center aligns the video at the top of the content, and if there isn't right floats it. We could update it to check for callouts too in addition to other media, now that these are getting into place.

**Comment by cory-skeers** [comment]
Similar to the featured video align, there are some places where an image and callout together start to overlap where on the original site they didn't.

https://now.uiowa.ddev.site/news/2015/03/improving-health-care-delivery
https://now.uiowa.edu/2015/04/improving-health-care-delivery

image

I'm not sure there's much we can do here to help it.

**Comment by cory-skeers** [comment]
> there are some places where an image and callout together start to overlap where on the original site they didn't

One option brought up during eod check-in that may work is adding `clear:right` to the callout. https://sandbox.prod.drupal.uiowa.edu/news/2023/03/ui-hosts-2015-ncaa-mens-swimming-and-diving-championships is available on sandbox as an example to play around with.

**Comment by cory-skeers** [comment]
> The logic that places the video in the body checks for any images within the first 700 characters of the body--if there is one, it center aligns the video at the top of the content, and if there isn't right floats it. We could update it to check for callouts too in addition to other media, now that these are getting into place.

This has been updated, and now checks for either media or callouts within the top section of the body.
image

**Comment by cory-skeers** [comment]
I think this would be a rabbit hole deeper than we'd want to go, but there are some places where they've used callouts which are a better fit for blockquotes:
https://now.uiowa.edu/2014/10/ui-dedicates-126-million-biomedical-hub

image

as opposed to 

image

But that would require something like some heavy regexing to identify callouts that are structured like blockquotes (assuming that over the years they haven't been consistently set up as `"Quote" - Name`

I think how these get set up and the layout for them is fine, just not optimal.

**Comment by cory-skeers** [comment]
There are some callouts where there are separate bold sections within the content, and these are getting placed in the callout heading
image

Original
image

https://now.uiowa.edu/2023/02/ui-named-top-producer-fulbright-us-students-seventh-year

**Comment by bspeare** [comment]
I think we should add the `clear: right` and `clear: left` properties to the `inline--align-right` and `inline--align-left` classes in UIDS.   All of our floated media `figure.align-right` and `figure.align-left` should probably have these defined as well in `uids_base`.   Does anyone have any concerns moving forward with this?

File: docroot/sites/now.uiowa.edu/modules/now_migrate/src/Plugin/migrate/source/Achievement.php [hunk 2]
```diff
@@ -203,11 +215,16 @@ private function calloutReplace($match) {
...
+    // as it is not in the same '<strong>' group
+    // as the ones on the first line.
+    $headline_match_string = preg_replace('%(<br>|<br \/>|<br\/>).*%is', '', $match[2]);
+
     // Look for a headline to use in the callout, which are bolded strings
     // at the start of the callout. Also look for any additional
     // line breaks. Like before, here they are unnecessary.
     $headline = '';
-    if (preg_match('|^<strong>(.*?)<\/strong>(<br>)*|is', $match[2], $headline_matches)) {
+    if (preg_match_all("|<strong>(.*?)<\/strong>(<br>)*|is", $headline_match_string, $headline_matches)) {
```
Comment (cory-skeers) [inline docroot/sites/now.uiowa.edu/modules/now_migrate/src/Plugin/migrate/source/Achievement.php]: Taking a look at this. It works great in most cases, but there's not always a `<br>` there, like in https://now.uiowa.edu/2023/02/ui-named-top-producer-fulbright-us-students-seventh-year. We can fix that case by adding `<ul>` to the regex matching options, but not sure if there might be other instances like this with other things.

**Comment by cory-skeers** [comment]
> I think we should add the `clear: right` and `clear: left` properties to the `inline--align-right` and `inline--align-left` classes in UIDS. All of our floated media `figure.align-right` and `figure.align-left` should probably have these defined as well in `uids_base`. Does anyone have any concerns moving forward with this?

Maybe limit it to the `inline--align-*` stuff for this work, and create a followup for the `figure.align-*` stuff in `uids_base`? I think it'd be a good change to make, but if it's only the inline stuff here, it doesn't add any more to testing (other than re-testing stuff) since callouts are the only inline items for now.

File: docroot/sites/now.uiowa.edu/modules/now_migrate/src/Plugin/migrate/source/Achievement.php (hunk 2 above)
Comment (cory-skeers) [reply docroot/sites/now.uiowa.edu/modules/now_migrate/src/Plugin/migrate/source/Achievement.php]: > We can fix that case by adding <ul> to the regex matching options...

Did this. Checking through to see if there are other cases.

**Comment by bspeare** [comment]
@cory-skeers that sounds like a good idea.  Do you want to make that change for the callouts in `uids_base` or `UIDS`?

**Comment by cory-skeers** [comment]
> that sounds like a good idea. Do you want to make that change for the callouts in `uids_base` or `UIDS`?

@bspeare I think adding to the inline alignment classes in `uids` now makes sense to me. It might be more to re-check when moving #6120 forward, but I think we need to go through that anyway, and it's probably something we want to add for that anyway.

**Comment by cory-skeers** [comment]
With https://github.com/uiowa/uids/pull/875, I think the overlapping issues like in https://github.com/uiowa/uiowa/pull/6280#issuecomment-1480255472 are no longer an issue (styling update, not migration update, so testing locally is easy if it's already been migrated).

I think with that, things look good to go. We'll want to take out the query limiters before merging, but otherwise 💯

**Comment by bspeare** [comment]
@cory-skeers I added this follow-up issue https://github.com/uiowa/uiowa/issues/6293 based on https://github.com/uiowa/uiowa/pull/6280#issuecomment-1481371332.

**Review by joewhitsitt — APPROVED** [review]
Approving based on standup conversation. Reviewed, not tested.

**Comment by cory-skeers** [comment]
Thanks @joewhitsitt!

**Pull request merged**